src/types/acls.rs

```diff
@@ -0,0 +1,287 @@
+use crate::error::{Error, ParseError};
+use crate::types::UnsolicitedResponse;
+use crate::parse::try_handle_unilateral;
+use imap_proto::Response;
+use std::collections::HashSet;
+use std::sync::mpsc;
+
+// re-export some types from imap_proto
+pub use imap_proto::types::AclRight;
+
+/// enum used for set_acl to specify how the ACL is to be modified.
+pub enum AclModifyMode {
+    /// Replace all ACLs on the identifier for the mailbox
+    Replace,
+    /// Add the given ACLs to the identifier for the mailbox
+    Add,
+    /// Remove the given ACLs from the identifier for the mailbox
+    Remove,
+}
+
+/// Helpful wrapper around the ACL rights vector
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclRightList {
+    pub(crate) data: HashSet<AclRight>,
+}
+
+impl AclRightList {
+    /// Returns if the AclRightList has the provided ACL (either as a char or an AclRight enum)
+    pub fn has_right<T: Into<AclRight>>(&self, right: T) -> bool {
+        self.data.contains(&right.into())
+    }
+}
+
+impl ToString for AclRightList {
+    fn to_string(&self) -> String {
+        let mut v: Vec<char> = self.data.iter().map(|c| char::from(*c)).collect();
+
+        v.sort_unstable();
+
+        v.into_iter().collect()
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRightList {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRightList {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRightList {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl From<&str> for AclRightList {
+    fn from(i: &str) -> Self {
+        i.chars()
+            .into_iter()
+            .map(|c| c.into())
```

Comment:
This seems dubious. Is _every_ `char` a valid `AclRight`? What happens if one of the characters is some super obscure Unicode code point?

Reply:
the spec only allows lower alpha characters to be defined as extra rights. And the way I built the AclRight enum is that the core-defined rights (in 4314 and the annotate right in 5257) are enumerated values in the enum.  And any other "char" is put under a "Custom(char)" enumeration.  In this way any custom right provided by the server (via RIGHTS= capability) will be handled correctly by this iteration without issue. (I can add another test to show that case)

Reply:
Right, but this doesn't check for "lower alpha characters". It takes _any_ Unicode symbol (`char`). I think we'll want to have this fail if the `&str` contains any non-lower-ASCII-alpha characters.

Reply:
ok.. I'll change it to a TryFrom.